x-pack/auditbeat/module/system/packages/packages.go

```diff
@@ -0,0 +1,336 @@
+	cfgwarn.Experimental("The %v/%v dataset is experimental", moduleName, metricsetName)
+
+	config := defaultConfig
+	if err := base.Module().UnpackConfig(&config); err != nil {
+		return nil, errors.Wrapf(err, "failed to unpack the %v/%v config", moduleName, metricsetName)
+	}
+
+	ms := &MetricSet{
+		BaseMetricSet: base,
+		config:        config,
+		log:           logp.NewLogger(moduleName),
+	}
+
+	if os, err := getOS(); err == nil {
+		switch os.Family {
+		case redhat, debian, darwin:
+			ms.osFamily = os.Family
+		default:
+			return nil, fmt.Errorf("this metricset does not support OS family %v", os.Family)
+		}
+	} else if err != nil {
+		return nil, err
+	}
+
+	if config.ReportChanges {
+		ms.cache = cache.New()
+	}
+
+	return ms, nil
+}
+
+// Fetch collects data about the host. It is invoked periodically.
+func (ms *MetricSet) Fetch(report mb.ReporterV2) {
+	packages, err := getPackages(ms.osFamily)
+	if err != nil {
+		ms.log.Error(err)
+		report.Error(err)
+		return
+	}
+
+	if ms.cache != nil && !ms.cache.IsEmpty() {
+		installed, removed := ms.cache.DiffAndUpdateCache(convertToCacheable(packages))
+
+		for _, pkgInfo := range installed {
+			pkgInfoMapStr := pkgInfo.(*Package).toMapStr()
+			pkgInfoMapStr.Put("status", "new")
+
+			report.Event(mb.Event{
+				MetricSetFields: common.MapStr{
+					"package": pkgInfoMapStr,
+				},
+			})
+		}
+
+		for _, pkgInfo := range removed {
+			pkgInfoMapStr := pkgInfo.(*Package).toMapStr()
+			pkgInfoMapStr.Put("status", "removed")
+
+			report.Event(mb.Event{
+				MetricSetFields: common.MapStr{
+					"package": pkgInfoMapStr,
+				},
+			})
+		}
+	} else {
+		// Report all installed packages
+		var pkgInfos []common.MapStr
+
+		for _, pkgInfo := range packages {
+			pkgInfoMapStr := pkgInfo.toMapStr()
+			pkgInfoMapStr.Put("status", "installed")
+
+			pkgInfos = append(pkgInfos, pkgInfoMapStr)
+		}
+
+		report.Event(mb.Event{
+			MetricSetFields: common.MapStr{
+				"package": pkgInfos,
+			},
+		})
+
+		if ms.cache != nil {
+			// This will initialize the cache with the current packages
+			ms.cache.DiffAndUpdateCache(convertToCacheable(packages))
+		}
+	}
+}
+
+func convertToCacheable(packages []*Package) []cache.Cacheable {
+	c := make([]cache.Cacheable, 0, len(packages))
+
+	for _, p := range packages {
+		c = append(c, p)
+	}
+
+	return c
+}
+
+func getPackages(osFamily string) (packages []*Package, err error) {
+	switch osFamily {
+	case redhat:
+		// TODO: Implement RPM
+		err = errors.New("RPM not yet supported")
+	case debian:
+		packages, err = listDebPackages()
+		if err != nil {
+			err = errors.Wrap(err, "error getting DEB packages")
+		}
+	case darwin:
+		packages, err = listBrewPackages()
+		if err != nil {
+			err = errors.Wrap(err, "error getting Homebrew packages")
+		}
+	default:
+		panic("unknown OS - this should not have happened")
+	}
+
+	return
+}
+
+func listDebPackages() ([]*Package, error) {
+	const statusFile = "/var/lib/dpkg/status"
```

Comment:
We should leave it for a follow up PR, but I think we should test this code by injecting the status file location and providing a sample.